okta/src/main/java/org/ccci/idm/user/okta/dao/OktaUserDao.kt

```diff
@@ -0,0 +1,354 @@
+import org.ccci.idm.user.okta.OktaGroup
+import org.ccci.idm.user.okta.dao.util.filterUsers
+import org.ccci.idm.user.okta.dao.util.oktaUserId
+import org.ccci.idm.user.okta.dao.util.searchUsers
+import org.ccci.idm.user.query.Attribute
+import org.ccci.idm.user.query.BooleanExpression
+import org.ccci.idm.user.query.ComparisonExpression
+import org.ccci.idm.user.query.Expression
+import org.ccci.idm.user.query.NotExpression
+import java.util.EnumSet
+import java.util.concurrent.BlockingQueue
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.stream.Stream
+
+private const val PROFILE_THEKEY_GUID = "theKeyGuid"
+private const val PROFILE_RELAY_GUID = "relayGuid"
+private const val PROFILE_EMAIL = "email"
+private const val PROFILE_FIRST_NAME = "firstName"
+private const val PROFILE_NICK_NAME = "nickName"
+private const val PROFILE_LAST_NAME = "lastName"
+
+private const val PROFILE_PHONE_NUMBER = "primaryPhone"
+private const val PROFILE_CITY = "city"
+private const val PROFILE_STATE = "state"
+private const val PROFILE_ZIP_CODE = "zipCode"
+private const val PROFILE_COUNTRY = "countryCode"
+
+private const val PROFILE_US_EMPLOYEE_ID = "usEmployeeId"
+private const val PROFILE_US_DESIGNATION = "usDesignationNumber"
+private const val PROFILE_EMAIL_ALIASES = "emailAliases"
+
+private val DEFAULT_ATTRS = arrayOf(User.Attr.EMAIL, User.Attr.NAME, User.Attr.FLAGS)
+
+class OktaUserDao(private val okta: Client, private val listeners: List<Listener>? = null) : AbstractUserDao() {
+    var maxSearchResults = SEARCH_NO_LIMIT
+    var initialGroups: Set<String> = emptySet()
+
+    private fun findOktaUser(user: User) =
+        findOktaUserByOktaUserId(user.oktaUserId) ?: findOktaUserByTheKeyGuid(user.theKeyGuid)
+
+    fun findByOktaUserId(id: String?) = findOktaUserByOktaUserId(id)?.asIdmUser()
+    private fun findOktaUserByOktaUserId(id: String?) = id?.let { okta.getUser(id) }
+
+    override fun findByEmail(email: String?, includeDeactivated: Boolean): User? {
+        if (email == null) return null
+        return okta.filterUsers("profile.email eq \"$email\"").firstOrNull()?.asIdmUser()
+    }
+
+    override fun findByTheKeyGuid(guid: String?, includeDeactivated: Boolean) =
+        findOktaUserByTheKeyGuid(guid)?.asIdmUser()
+    private fun findOktaUserByTheKeyGuid(guid: String?) =
+        guid?.let { okta.searchUsers("profile.$PROFILE_THEKEY_GUID eq \"$guid\"").firstOrNull() }
+
+    override fun findByRelayGuid(guid: String?, includeDeactivated: Boolean): User? {
+        if (guid == null) return null
+        return okta.searchUsers("profile.$PROFILE_RELAY_GUID eq \"$guid\"").firstOrNull()?.asIdmUser()
+    }
+
+    // region Stream Users
+    override fun streamUsers(
+        expression: Expression?,
+        includeDeactivated: Boolean,
+        restrictMaxAllowed: Boolean
+    ): Stream<User> {
+        val search = expression?.toOktaExpression()
+        return okta.listUsers(null, null, null, search, null).stream()
```

Comment:
really neat that the SDK can return a stream for us

Reply:
Yeah, it automatically handles paging which is really nice